src/coreComponents/constitutive/fluid/multifluid/CO2Brine/functions/CO2Solubility.hpp

```diff
@@ -229,112 +241,169 @@ CO2SolubilityUpdate::compute( real64 const & pressure,
+    LvArray::forValuesInSlice( phaseFraction.value, setZero );
+    LvArray::forValuesInSlice( phaseCompFraction.value, setZero );
+    LvArray::forValuesInSlice( phaseFraction.derivs, setZero );
+    LvArray::forValuesInSlice( phaseCompFraction.derivs, setZero );
 
-    phaseFraction.value[m_phaseGasIndex] = 1.0 - phaseFraction.value[m_phaseLiquidIndex];
+    // 1) Compute phase fractions
 
-    phaseFraction.derivs[m_phaseGasIndex][Deriv::dP] = -phaseFraction.derivs[m_phaseLiquidIndex][Deriv::dP];
-    phaseFraction.derivs[m_phaseGasIndex][Deriv::dT] = -phaseFraction.derivs[m_phaseLiquidIndex][Deriv::dT];
-    phaseFraction.derivs[m_phaseGasIndex][Deriv::dC+m_CO2Index] = -phaseFraction.derivs[m_phaseLiquidIndex][Deriv::dC+m_CO2Index];
-    phaseFraction.derivs[m_phaseGasIndex][Deriv::dC+m_waterIndex] = -phaseFraction.derivs[m_phaseLiquidIndex][Deriv::dC+m_waterIndex];
+    phaseFraction.value[activePhase] = 1.0;
 
     // 2) Compute phase component fractions
-
-    // liquid phase composition  CO2 = Csat / (Csat + W) = (Csat/W) / (Csat/W + 1)
-    real64 const onePlusSolubilityInv = 1.0 / ( 1.0 + solubility );
-    phaseCompFraction.value[m_phaseLiquidIndex][m_CO2Index] = solubility * onePlusSolubilityInv;
-
-    phaseCompFraction.derivs[m_phaseLiquidIndex][m_CO2Index][Deriv::dP] = solubilityDeriv[0] * (onePlusSolubilityInv*onePlusSolubilityInv);
-    phaseCompFraction.derivs[m_phaseLiquidIndex][m_CO2Index][Deriv::dT] = solubilityDeriv[1] * (onePlusSolubilityInv*onePlusSolubilityInv);
-
-    phaseCompFraction.value[m_phaseLiquidIndex][m_waterIndex] = 1.0 - phaseCompFraction.value[m_phaseLiquidIndex][m_CO2Index];
-
-    phaseCompFraction.derivs[m_phaseLiquidIndex][m_waterIndex][Deriv::dP] = -phaseCompFraction.derivs[m_phaseLiquidIndex][m_CO2Index][Deriv::dP];
-    phaseCompFraction.derivs[m_phaseLiquidIndex][m_waterIndex][Deriv::dT] = -phaseCompFraction.derivs[m_phaseLiquidIndex][m_CO2Index][Deriv::dT];
-
-    // gas phase composition  CO2 = 1.0
-
-    phaseCompFraction.value[m_phaseGasIndex][m_CO2Index]   = 1.0;
-    phaseCompFraction.value[m_phaseGasIndex][m_waterIndex] = 0.0;
-
-    phaseCompFraction.derivs[m_phaseGasIndex][m_CO2Index][Deriv::dP]   = 0.0;
-    phaseCompFraction.derivs[m_phaseGasIndex][m_waterIndex][Deriv::dT] = 0.0;
-    phaseCompFraction.derivs[m_phaseGasIndex][m_CO2Index][Deriv::dP]   = 0.0;
-    phaseCompFraction.derivs[m_phaseGasIndex][m_waterIndex][Deriv::dT] = 0.0;
-    // phaseCompFraction does not depend on globalComponentFraction
-
+    // Setup default values which will be overridden for the active phase
+    phaseCompFraction.value[m_phaseGasIndex][m_CO2Index] = 1.0;
+    phaseCompFraction.value[m_phaseLiquidIndex][m_waterIndex] = 1.0;
+    // Set the global composition as the composition of the active phase
+    for( integer ic = 0; ic < numComps; ++ic )
+    {
+      phaseCompFraction.value[activePhase][ic] = compFraction[ic];
+      phaseCompFraction.derivs[activePhase][ic][Deriv::dC+ic] = 1.0;
```

Comment:
Is this for sure ?

Reply:
Yes. Again valid as long as we remain single phase.